### ISSUE TITLE
Support for Re-Use of Reduction Instances

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#545]
It would be nice if we could re-use reduction instances both for multiples tasks within the same reduction epoch as well as across different reduction epochs. There are slightly different analyses needed to support this, but they are relatively similar. There should be a way to leverage out knowledge about the mapping dependence process to enable this. We do have to be a little bit careful about how this works across multiple different contexts with non-leaf tasks that have only reduction privileges.

**Comment by mariodirenzo** [comment]
This feature would be very useful for an application like this
```
import "regent"

local C = regentlib.c
local cos  = regentlib.cos(double)

local Np = 10
local Nx = 2
local nStep = 2000

struct Fields { x : double }

__demand(__cuda, __leaf, __parallel)
task updateMsh(msh : region(ispace(int2d),Fields),
               i : int)
where
   writes(msh.x)
do
   __demand(__openmp)
   for c in msh do
      msh[c].x = cos(double(i+int2d(c).x)/nStep*2.0*3.14)
   end
end

__demand(__cuda, __leaf)
task addAvg(msh : region(ispace(int2d),Fields),
            avg : region(ispace(int1d),Fields),
            fact : double)
where
   reads(msh.x),
   reduces+(avg.x)
do
   __demand(__openmp)
   for c in msh do
      avg[c.x].x += msh[c].x*fact
   end
end

__demand(__leaf)
task outputAvg(avg : region(ispace(int1d),Fields) )
where
   reads(avg.x)
do
   C.printf("Stats:\n")
   for c in avg do
      C.printf("%d\t%10f\n", c, avg[c].x)
   end
   C.printf("\n")
   C.printf("\n")
end

task main()

   var msh = region(ispace(int2d,{Np, Np}),Fields)
   var avg = region(ispace(int1d,Np),Fields)

   var is_msh = ispace(int2d, {Nx, Nx})
   var is_avg = ispace(int1d, Nx)

   var p_msh = partition(equal,msh,is_msh)
   var p_avg = partition(equal,avg,is_avg)

   fill(msh.x, 0.0)
   fill(avg.x, 0.0)

   for i=0,nStep do
      updateMsh(msh, i)
      __demand(__parallel)
      for c in is_msh do
         addAvg(p_msh[c], p_avg[c.x], 1.0/nStep)
      end
   end

   outputAvg(avg)
end

regentlib.start(main)
```
My understanding is that legion performs ``nStep`` reductions right now.
They could be reduced to just one.

**Comment by lightsighter** [comment]
This issue has gotten tied up with my refactoring of the internals of instances in the `newinsts` branch. I expect that it will be considerably easier to implement this once that refactoring is complete. Moving this to the 20.06 milestone.

**Comment by lightsighter** [comment]
Work on this issue is being tracked with this [merge request](https://gitlab.com/StanfordLegion/legion/-/merge_requests/183)

**Comment by lightsighter** [comment]
There is support for this issue now in the `newinsts` branch. It will likely merge into the `master` branch not long after the 20.06 release. The default mapper has been updated to reuse reduction instances as well.

The one case that we have not fully handled is the ABA reduction case, where an application first performs reductions of kind A to a field, then performs reductions of kind B, and then goes back to reductions of kind A. If the mapper chooses to reuse a reduction instance of for kind A in both the first and last epochs of reduction A operations, then that could lead to wrong answers because reductions will not be properly ordered with respect to the B reductions. The runtime currently will detect this case and report it every time that it happens, but it there is no mechanism at the moment for the mapper to check to see if this case is happening when it is making its mapping decision. I believe that this case is exceedingly rare and the error message for the detection explicitly asks the user to email the legion developer's mailing list when it occurs, so until we see it in the wild then I'm going to leave it alone for now.

**Comment by lightsighter** [comment]
Support for this is now in the `master` branch of Legion and has been upstreamed to the `control_replication` branch.
